### PR TITLE
analysis: add a package for static analysis

### PR DESCRIPTION
**Pull request by pgavlin**
These changes add a package to aid in authoring tools that use the results of Check/Open as a jumping-off point for analyzing environments. This package currently supports finding the expression at a given source position and describing the entity at a given source position.

### CODE REVIEW

**Comment by pgavlin** [comment]
ping.

File: expr.go [hunk 1]
```diff
@@ -29,22 +29,25 @@ type Expr struct {
...
+	KeyRanges map[string]Range `json:"keyRanges,omitempty"`
+
 	// The fields below act as a discriminated union. Only one must be non-nil at any given time. If all fields are nil,
 	// then this Expr is a null literal expression.
 
 	// The literal value, if this is a literal expression (nil, bool, json.Number, or string)
 	Literal any `json:"literal,omitempty"`
 
-	// The interpolations, if this is a a string interpolation expression
+	// The interpolations, if this is a a string interpolation expression.
```
Comment (komalali) [inline expr.go:41]:
```suggestion
	// The interpolations, if this is a string interpolation expression.
```

File: expr.go [hunk 2]
```diff
@@ -29,22 +29,25 @@ type Expr struct {
...
 	Interpolate []Interpolation `json:"interpolate,omitempty"`
 
-	// The property accessors, if this is a a symbol expression
+	// The property accessors, if this is a a symbol expression.
```
Comment (komalali) [inline expr.go:44]:
```suggestion
	// The property accessors, if this is a symbol expression.
```

File: expr.go [hunk 3]
```diff
@@ -29,22 +29,25 @@ type Expr struct {
...
 	Symbol []PropertyAccessor `json:"symbol,omitempty"`
 
-	// The access, if this is an access expression
+	// The access, if this is an access expression.
 	Access *AccessExpr `json:"access,omitempty"`
 
-	// The list elements, if this is a list expression
+	// The list elements, if this is a list expression.
 	List []Expr `json:"list,omitempty"`
```
Comment (komalali) [inline expr.go:51]: I thought we renamed `List` to `Array`?

File: analysis/describe.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+func (a *Analysis) describeBuiltin(builtin *esc.BuiltinExpr) (string, bool) {
+	switch builtin.Name {
+	case "fn::fromJSON":
+		return "Decodes a value from its JSON representation.", true
+	case "fn::fromBase64":
+		return "Decodes a string from its Base64 representation.", true
+	case "fn::join":
+		return "Concatenates the elements of its second argument to create a single string. The second argument is" +
+			"places between each elements in the result.", true
```
Comment (komalali) [inline analysis/describe.go:63]:
```suggestion
		return "Concatenates the elements of its second argument to create a single string. The first argument is " +
			"placed between each element in the result.", true
```

File: analysis/describe_test.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	require.Empty(t, diags)
+
+	analysis := New(*env, map[string]*schema.Schema{"test": testProviderSchema})
+
+	expected := map[esc.Pos]string{
+		{Line: 5, Column: 5}:   "Decodes a string from its Base64 representation.",
+		{Line: 5, Column: 21}:  "string",
+		{Line: 7, Column: 5}:   "Decodes a value from its JSON representation.",
+		{Line: 7, Column: 19}:  "string",
+		{Line: 9, Column: 5}:   "Concatenates the elements of its second argument to create a single string. The second argument isplaces between each elements in the result.",
```
Comment (komalali) [inline analysis/describe_test.go:44]:
```suggestion
		{Line: 9, Column: 5}:   "Concatenates the elements of its second argument to create a single string. The first argument is placed between each element in the result.",
```

File: analysis/traversal.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+// ExpressionAtPos returns the expression that contains pos. If no such expression exists, expressionAtPos returns
+// (nil, false). The returned expression is the "smallest" expression that contains pos.
+func (a *Analysis) ExpressionAtPos(pos esc.Pos) (*esc.Expr, bool) {
+	x, _, ok := a.expressionAtPos(pos)
+	return x, ok
+}
+
+// expressionAtPos returns the expression that contains pos. If no such expression exists, expressionAtPos returns
+// (nil, nil, false). The returned expression is the "smallest" expression that contains pos. The path to the given
+// is also returned.
```
Comment (komalali) [inline analysis/traversal.go:47]:
```suggestion
// expression is also returned.
```

**Review by komalali — APPROVED** [review]
LGTM, mostly some nits and one necessary fix for the description of fn::Join

File: expr.go (hunk 3 above)
Comment (pgavlin) [reply expr.go:51]: Still need to do so in the public API :(

Comment (pgavlin) [reply expr.go:51]: (I deliberately left this part out for the moment)

**Pull request merged**